### ISSUE TITLE
Fixed-point version for FPU-less systems

### ISSUE THREAD

**Issue by Cacodemon345** [schellingb/TinySoundFont#89]
A fixed-point version of the library would be nice for systems with no FPU (e.g. 32-bit ARM systems with enough memory to handle reasonable soundfonts) and also for purposes of using it in the likes of SBEMU/VSBHDA (where reliance on the FPU can cause problems).

**Comment by viti95** [comment]
This should benefit systems with weak FPUs, such as the Vortex86 or VIA CPUs. For example, a 933 MHz Vortex86MX has FPU performance comparable to a Pentium II 350 MHz and is incapable of handling real-time MIDI synthesis:

https://youtu.be/klvMr7NKQq8

**Comment by alexriegler12** [comment]
Yes, it would be nice to do a fixed point version (for eg raspberry pi pico) and also make it independent of little endian.

**Comment by earlephilhower** [comment]
FWIW I've had a (mostly) fixed point port of TinySoundFont for microcontrollers like the ESP8266, ESP32-family, and Raspberry Pi Pico-family in [ESP8266Audio](https://github.com/earlephilhower/ESP8266Audio).

As of now there are still floating point operations in the envelope calculations (because getting the dynamic range right is a bear, at least for me).  The latest version just converted SF2 loading( to use ROM arrays directly (which should work on the R.Pi SBCs just be in RAM) since the total RAM for these microcontrollers can be as low as 40KB of heap.  On the RP2040 Pico it handles relatively complex MIDs at 22KHz stereo in realtime.  On the RP2350 (which has a single precision FPU) it's good enough for 44.1khz on many moderately complex MIDs.

**Comment by alexriegler12** [comment]
Maybe it would bei good to create a standalone Version of the esp8266audio tinysoundfont Version that does not depend on Arduino libraries.

**Comment by earlephilhower** [comment]
I'm actually working off of a fork of the library, so you can look at https://github.com/earlephilhower/TinySoundFont which should build w/o the Arduino infra.  There are some add'l defines to enable the FixedPt mode and steps to make the header-version of the SoundFont (so it remains in ROM/flash on the embedded systems and doesn't take up precious SRAM) but all the tools and host-side tests are there in the linux-xxx examples.  Whether the optimizations done to make it fork on the 200Mhz ARM M33 and Xtensa chips make sense for the multi-GHz R.Pis, I really don't know.  (i.e. the approximation of `powf` patched into the code bumped up performance a whopping 10% on those embedded chips because the transcendental ops are SW-emulated and a few ULP of precision aren't important there)

**Comment by alexriegler12** [comment]
I have successfully played midi files on the Nintendo DS using the fixed point fork from the arduino library. Maybe I will test it on RP2040 or RP2350
